src/Runner.Worker/Handlers/ScriptStalledManager.cs

```diff
@@ -0,0 +1,64 @@
+using System;
+using System.Collections.Generic;
+using System.Timers;
+using GitHub.Runner.Common;
+using GitHub.Runner.Sdk;
+using GitHub.Runner.Worker;
+
+namespace GitHub.Runner.Worker.Handlers
+{
+    [ServiceLocator(Default = typeof(TimerAdapter))]
+    public interface ITimer 
+    {
+        void Start();
+        void Stop();
+        double Interval { get; set; }
+        event ElapsedEventHandler Elapsed;
+        bool AutoReset { get; set; }
+        void Dispose();
+    }
+
+    public class TimerAdapter : Timer, ITimer {}
+
+    public sealed class ScriptStalledManager : IDisposable
+    {
+        private readonly IExecutionContext _executionContext;
+        private readonly double _interval;
+
+        private ITimer _timer { get; set; }
+        public int IntervalsElapsedWhileStalled { get; private set;}
+
+        public ScriptStalledManager(IExecutionContext executionContext, double interval, ITimer timer)
+        {
+            _executionContext = executionContext;
+            _interval = interval;
+            _timer = timer;
+
+            IntervalsElapsedWhileStalled = 0;
+
+            _timer.Interval = _interval;
+            _timer.Elapsed += TriggerWarning;
+        }
+
+        public ScriptStalledManager(IExecutionContext executionContext, double interval): this(executionContext, interval, new TimerAdapter()) {}
+
+        public void Dispose()
+        {
+            _timer.Dispose();
+        }
+
+        public void OnDataReceived(object sender, ProcessDataReceivedEventArgs e)
+        {
+            IntervalsElapsedWhileStalled = 0;
+            _timer.Stop();
+            _timer.Start();
+        }
+
+        private void TriggerWarning(object source, ElapsedEventArgs e)
+        {
+            IntervalsElapsedWhileStalled++;
+            string friendlyDuration = StringUtil.TimespanToHumanReadableString(TimeSpan.FromMilliseconds(IntervalsElapsedWhileStalled * _interval));
+            _executionContext.Warning($"No output has been detected in the last {friendlyDuration} and the process has not yet exited. This step may have stalled and might require some investigation.");
```

Comment:
we might want to consult with PM to see what other information we should/can provide to make this more actionable to our customers.